kernel/fatal.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <sys/printk.h>
+#include <sys/__assert.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <logging/log.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
+{
+	ARG_UNUSED(reason);
+
+	/* TODO: What's the best way to totally halt the system if SMP
+	 * is enabled?
+	 */
+
+	(void)z_arch_irq_lock();
+	for (;;) {
+		k_cpu_idle();
+	}
+}
+/* LCOV_EXCL_STOP */
+
+/* LCOV_EXCL_START */
+__weak void k_sys_fatal_error_handler(unsigned int reason,
+				      const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
+	LOG_PANIC();
+	z_fatal_print("Halting system");
+	z_arch_system_halt(reason);
+	CODE_UNREACHABLE;
+}
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_LOG) || defined(CONFIG_PRINTK)
+static const char *thread_name_get(struct k_thread *thread)
+{
+	const char *thread_name = k_thread_name_get(thread);
+
+	if (thread_name == NULL || thread_name[0] == '\0') {
+		thread_name = "unknown";
+	}
+
+	return thread_name;
+}
+
+static const char *reason_to_str(unsigned int reason)
+{
+	switch (reason) {
+	case K_ERR_CPU_EXCEPTION:
+		return "CPU exception";
+	case K_ERR_SPURIOUS_IRQ:
+		return "Unhandled interrupt";
+	case K_ERR_STACK_CHK_FAIL:
+		return "Stack overflow";
+	case K_ERR_KERNEL_OOPS:
+		return "Kernel oops";
+	case K_ERR_KERNEL_PANIC:
+		return "Kernel panic";
+	default:
+		return "Unknown error";
+	}
+}
+
+void z_fatal_print(const char *fmt, ...)
```

Comment:
what about changing it to macro like:
```
#define z_fatal_print(fmt,...) \
    if (IS_ENABLED(CONFIG_LOG)) LOG_ERR(fmt,__VA_ARGS__); \
    else if (IS_ENABLED(CONFIG_PRINTK)) printk("FATAL: " fmt "\n", __VA_ARGS__) 
```
this macro reduces number of `#if defined(CONFIG_LOG) || defined(CONFIG_PRINTK)` as compiler will take care of what is used and what unused.

note that there is no `LOG_PANIC` there as it should be called explicitly at the place where decision about recovering is made.

Reply:
I already told you. I can't put this in the header. It results in an #include dependency loop. Fixing that would make this patch even more invasive than it already is. We need to move on.

Reply:
I removed the LOG_PANIC() call, that wasn't supposed to be there, it gets called in k_sys_fatal_error_handler() only.

Reply:
Also please note: the ifdefs are still necessary. gc-sections doesn't work the way you might expect and the string literals are not discarded even if the functions are unused. See https://stackoverflow.com/questions/54996229/is-ffunction-sections-fdata-sections-and-gc-sections-not-working

Reply:
hm, that i didn't know. However, I've did simple test in hello world and it seems like they are removed.

> I already told you

Sorry, i missed that. Yes, macro may result in include looping :/ hopefully we can fix it later.

Reply:
created #17841 which should cut the loop.